From f14f69d996ffc1c61dd523b839271d4a51328e4d Mon Sep 17 00:00:00 2001
From: Luca Boccassi <bluca@debian.org>
Date: Fri, 1 Jan 2021 19:04:37 +0000
Subject: [PATCH 1/3] cmake: link dynamically to libclang-cpp if found and
 ENABLE_LLVM_SHARED is set

ENABLE_LLVM_SHARED allows to dynamically link against libLLVM, but
libclang is still unconditionally linked statically.
Search for libclang-cpp.so, and if it is found and ENABLE_LLVM_SHARED
is set dynamically link against it.
Also expand the libstdc++ static linking check to include this new
condition.
---
 CMakeLists.txt               | 1 +
 cmake/clang_libs.cmake       | 4 ++++
 cmake/static_libstdc++.cmake | 2 +-
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 74fe4f19..b2e334e9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -76,6 +76,7 @@ find_library(libclangRewrite NAMES clangRewrite clang-cpp HINTS ${CLANG_SEARCH})
 find_library(libclangSema NAMES clangSema clang-cpp HINTS ${CLANG_SEARCH})
 find_library(libclangSerialization NAMES clangSerialization clang-cpp HINTS ${CLANG_SEARCH})
 find_library(libclangASTMatchers NAMES clangASTMatchers clang-cpp HINTS ${CLANG_SEARCH})
+find_library(libclang-shared libclang-cpp.so HINTS ${CLANG_SEARCH})
 if(libclangBasic STREQUAL "libclangBasic-NOTFOUND")
   message(FATAL_ERROR "Unable to find clang libraries")
 endif()
diff --git a/cmake/clang_libs.cmake b/cmake/clang_libs.cmake
index c33b635c..3f1523b7 100644
--- a/cmake/clang_libs.cmake
+++ b/cmake/clang_libs.cmake
@@ -26,6 +26,9 @@ llvm_map_components_to_libnames(_llvm_libs ${llvm_raw_libs})
 llvm_expand_dependencies(llvm_libs ${_llvm_libs})
 endif()
 
+if(ENABLE_LLVM_SHARED AND NOT libclang-shared STREQUAL "libclang-shared-NOTFOUND")
+set(clang_libs ${libclang-shared})
+else()
 # order is important
 set(clang_libs
   ${libclangFrontend}
@@ -46,6 +49,7 @@ list(APPEND clang_libs
   ${libclangAST}
   ${libclangLex}
   ${libclangBasic})
+endif()
 
 # prune unused llvm static library stuff when linking into the new .so
 set(_exclude_flags)
diff --git a/cmake/static_libstdc++.cmake b/cmake/static_libstdc++.cmake
index 3c8ac179..787ed9ad 100644
--- a/cmake/static_libstdc++.cmake
+++ b/cmake/static_libstdc++.cmake
@@ -1,7 +1,7 @@
 # only turn on static-libstdc++ if also linking statically against clang
 string(REGEX MATCH ".*[.]a$" LIBCLANG_ISSTATIC "${libclangBasic}")
 # if gcc 4.9 or higher is used, static libstdc++ is a good option
-if (CMAKE_COMPILER_IS_GNUCC AND LIBCLANG_ISSTATIC)
+if (CMAKE_COMPILER_IS_GNUCC AND LIBCLANG_ISSTATIC AND (NOT ENABLE_LLVM_SHARED OR libclang-shared STREQUAL "libclang-shared-NOTFOUND"))
   execute_process(COMMAND ${CMAKE_C_COMPILER} -dumpversion OUTPUT_VARIABLE GCC_VERSION)
   if (GCC_VERSION VERSION_GREATER 4.9 OR GCC_VERSION VERSION_EQUAL 4.9)
     execute_process(COMMAND ${CMAKE_C_COMPILER} -print-libgcc-file-name OUTPUT_VARIABLE GCC_LIB)
-- 
2.29.2

